### PR TITLE
feat: add sov-modules-core

### PR DESCRIPTION
**Pull request by vlopes11**
This commit introduces `sov-modules-core`, a new workspace member that will contain the required declarations for `no-std` `DispatchCall` runtime message declarations.

The new module is re-exported via `sov-state` and `sov-modules-api` to minimize breaking changes downstream.

Finally, it introduces a `RefCount` in `sov-rollup-interface` that will contain atomic synchronization, if the compile target contains atomic pointers. It safely fallback to `Rc`, that is a reference count without atomic synchronization, but equivalent API, that will be the used type on targets without synchronization. This can replace all the `Arc` instances in the future, but for now, it is required only for the previously `sov-state` storage/cache keys and values, so the `Storage` trait can be properly defined.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -25,10 +25,10 @@ members = [
     "module-system/sov-modules-stf-template",
     "module-system/sov-modules-rollup-template",
     "module-system/sov-modules-macros",
+    "module-system/sov-modules-core",
     "module-system/sov-state",
     "module-system/sov-modules-api",
     "module-system/module-schemas",
-    "module-system/utils/sov-first-read-last-write-cache",
```
Comment (citizen-stig) [inline Cargo.toml:34]: Do I understand corrrectly, that this crate merged into `sov-modules-core`

File: adapters/celestia/Cargo.toml [hunk 2]
```diff
@@ -9,7 +9,7 @@ publish = false
 
 [dependencies]
 borsh = { workspace = true, features = ["bytes"] }
-bech32 = { workspace = true }
+bech32 = { workspace = true, default-features = true }
```
Comment (citizen-stig) [inline adapters/celestia/Cargo.toml:12]: I honestly don't know, but what is the point of having `default-features = true` explicitly

File: module-system/sov-modules-core/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+name = "sov-modules-core"
+description = "Defines the core components of the Sovereign SDK module system"
+authors = { workspace = true }
+edition = { workspace = true }
+homepage = { workspace = true }
+license = { workspace = true }
+repository = { workspace = true }
+
+version = { workspace = true }
+readme = "README.md"
```
Comment (citizen-stig) [inline module-system/sov-modules-core/Cargo.toml:11]: It looks like REDME file is missing

File: module-system/sov-modules-core/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (citizen-stig) [inline module-system/sov-modules-core/src/lib.rs:1]: Can we also have `#![deny(missing_docs)]` here, so we don't have to do it again.

File: module-system/sov-state/src/codec/borsh_codec.rs [hunk 5]
```diff
@@ -7,7 +10,7 @@ pub struct BorshCodec;
 
 impl<K> StateKeyCodec<K> for BorshCodec
 where
-    K: borsh::BorshSerialize + borsh::BorshDeserialize,
+    K: BorshSerialize + borsh::BorshDeserialize,
```
Comment (citizen-stig) [inline module-system/sov-state/src/codec/borsh_codec.rs]: Nit: could we stick to one approach for both?

**Review by citizen-stig — APPROVED** [review]
This is big change, hard to spot if something is missing.

But LGTM, only couple small comments.

File: module-system/sov-modules-core/src/storage.rs [hunk 6]
```diff
@@ -1,24 +1,26 @@
-use std::fmt::Display;
```
Comment (bkolad) [inline module-system/sov-modules-core/src/storage.rs]: The file hierarchy is too flat, and we should split it logically into several rust modules:
- `storage, witness, cache, scratchpad` 
- `key, codec`
- everything else

File: module-system/sov-modules-api/src/dispatch.rs [hunk 7]
```diff
@@ -43,10 +23,27 @@ pub trait EncodeCall<M: Module> {
...
-pub trait CliWallet: DispatchCall {
+pub trait CliWallet: sov_modules_core::DispatchCall {
     /// The type that is used to represent this type in the CLI. Typically,
     /// this type implements the clap::Subcommand trait. This type is generic to
     /// allow for different representations of the same type in the interface; a
     /// typical end-usage will impl traits only in the case where `CliStringRepr<T>: Into::RuntimeCall`
     type CliStringRepr<T>;
 }
+
+impl<T> Genesis for T
```
Comment (bkolad) [inline module-system/sov-modules-api/src/dispatch.rs]: What is the methodology behind moving code from `sov-modules-api` to `core`? 
If we moved the `Module and DispatchCall` traits there, we should also move `Genesis`, so they are next to each other.

File: Cargo.toml (hunk 1 above)
Comment (vlopes11) [reply Cargo.toml:34]: Yes, that is correct. All the types there are required by `sov-modules-core`

File: adapters/celestia/Cargo.toml (hunk 2 above)
Comment (vlopes11) [reply adapters/celestia/Cargo.toml:12]: `bech32` is a requirement of `sov-modules-core`, and it must be `default-features = false`. This way, the bare minimum we declare on the root workspace is with such state, and if the library is declared elsewhere, it is expected to have the default features as these crates won't split std/no-std by feature

File: module-system/sov-modules-core/Cargo.toml (hunk 3 above)
Comment (vlopes11) [reply module-system/sov-modules-core/Cargo.toml:11]: Good catch; fixed

File: module-system/sov-modules-core/src/lib.rs (hunk 4 above)
Comment (vlopes11) [reply module-system/sov-modules-core/src/lib.rs:1]: Good suggestion; updated

File: module-system/sov-state/src/codec/borsh_codec.rs (hunk 5 above)
Comment (vlopes11) [reply module-system/sov-state/src/codec/borsh_codec.rs]: Done!

File: module-system/sov-modules-api/src/dispatch.rs (hunk 7 above)
Comment (vlopes11) [reply module-system/sov-modules-api/src/dispatch.rs]: The approach was to move only the minimum possible set of traits/structs to core so we can declare a runtime as no-std, but it makes sense. Done!

File: module-system/sov-modules-core/src/storage.rs (hunk 6 above)
Comment (vlopes11) [reply module-system/sov-modules-core/src/storage.rs]: Good suggestion; divided into 2 levels

**Review by bkolad — APPROVED** [review]

**Pull request merged**